File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Utility/LoseBlockAction.ts
import { FairyGUI, UnityEngine } from "csharp";
import Async from "../../../../../Core/Async";
import DungeonManager from "../../DungeonManager";
import { HpBlockBrokenEffect } from "../../Effect/HpBlockBrokenEffect";
import AbstractCreature from "../../Unit/AbstractCreature";
import AbstractGameAction, { ActionType } from "../AbstractGameAction";

export default class LoseBlockAction extends AbstractGameAction
{
    private Target:AbstractCreature;
    private Source:AbstractCreature;
    private Amount:number;
    private NoAnimation:boolean;
    Type = ActionType.BLOCK;
    public constructor(target : AbstractCreature, amount : number,noAnimation:boolean = false, source: AbstractCreature = null) {
        super();
        this.Target = target;
        this.NoAnimation = noAnimation;
        if(!source)
            this.Source = source;
        else
            this.Source = target;
        this.Amount = amount;
    }

    async Update(){
        if(this.Target && !this.Target.IsDying && !this.Target.IsDead){
            return true;
        }
        if(this.Amount == 0)
            return true;
        if(this.Target.Block == 0)
            return true;
        this.Target.Block -= this.Amount;
        if(this.Target.Block < 0)
            this.Target.Block = 0;
        if(this.Target.Block == 0){
            if(!this.NoAnimation){
                let hitbox = this.Target.GetHitBox();
                DungeonManager.EffectManager.Play(new HpBlockBrokenEffect(hitbox.CX,hitbox.CY))
            }
        } else {
            let blockCom = <FairyGUI.GComponent>this.Target.HealthComponent.GetChild("Block");
            blockCom.SetScale(5,5);
            blockCom.TweenScale(UnityEngine.Vector2.one,0.7).SetDelay(0.3);
        }
        await Async.Delay(500);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Combat/LoseBlockAction.cs
using Cysharp.Threading.Tasks;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect.Combat;
using ModTheSpire.Event;
using ModTheSpire.Unit;

namespace ModTheSpire.Effect.Other
{
    public class LoseBlockAction : IGameAction
    {
        public float Duration { get; set; }
        public ActionType Type { get; } = ActionType.BLOCK;
        private AbstractCreature mTarget;
        private int mAmount;
        private bool mNoAnimation;
        private AbstractCreature mSource;
        public void Init()
        {
        }

        public LoseBlockAction(AbstractCreature target, int amount, bool noAnimation = false, AbstractCreature source = null)
        {
            mTarget = target;
            mAmount = amount;
            mNoAnimation = noAnimation;
            mSource = source ?? target;
        }

        public async UniTask<bool> Update()
        {
            if(mTarget is { IsDying: false, IsDead: false }){
                return true;
            }
            if(mAmount == 0)
                return true;
            if(mTarget.Block == 0)
                return true;
            mTarget.Block -= mAmount;
            if(mTarget.Block < 0)
                mTarget.Block = 0;
            if(mTarget.Block == 0){
                if(!mNoAnimation){
                    var hitbox = mTarget.DisplayObject.GetHitBox();
                    DungeonManager.EffectManager.Play(new HpBlockBrokenEffect(hitbox.center.x, hitbox.center.y));
                }
            } else {
                var blockCom = mTarget.DisplayObject.UnitComponent.GetChild("Block");
                blockCom.SetScale(5,5);
                blockCom.TweenScale(UnityEngine.Vector2.one,0.7f).SetDelay(0.3f);
            }
            await UniTask.Delay(500);
            return true;
        }
    }
}